projecttacoma/deqm-test-server
# Validation Env

# Summary
DEQM test server now supports optional FHIR validation through Inferno's validation server!

## New behavior
When the VALIDATE, VALIDATION_HOST, and VALIDATION_URL are all set appropriately, DEQM test server will now query the Inferno FHIR validation server to ensure the bodies of all POST and PUT requests which attempt to add to or alter data on the server are valid FHIR. Also updates the meta.profile of the submitted resource to contain all profiles validated against

## Code changes
- Added validation middleware
- Added resourceValidationUtils file
- validationUtils.js -> operationValidationUtils.js
- Changed server initialize function to async
- Updated tests to use new async initialize function
- Added testing for resourceValidationUtils

# Testing guidance
- `npm run test`
- run the Inferno FHIR validator with `docker run -p 4567:4567 infernocommunity/fhir-validator-service`
- In a new terminal, start up the test server using `VALIDATE=true VALIDATION_HOST=localhost VALIDATION_PORT=4567 npm run start`
- Post/Put resources to the server, both valid and invalid. Ensure that the errors you get are intuitive and the meta.profile of the submitted resource is updated to include the base resource type (i.e. Patient, Observation... etc) as well as the qicore resource type when the resource is valid against qicore
- Try using resources from connectathon bundles, especially those with included meta.profiles and ensure that the validator validates against those profiles as well. NOTE: most of these resources (if not all) are actually invalid and will fail.
- If you have issues finding resources to try, just ping me! Happy to show you what I used.
- Run measure operations as well as GET and DELETE requests and ensure they are unaffected.
